src/coreComponents/linearAlgebra/DofManager.cpp

```diff
@@ -175,7 +176,107 @@ void forMeshSupport( std::vector< DofManager::FieldSupport > const & support,
 
 } // namespace
 
-void DofManager::createIndexArray( FieldDescription const & field )
+
+void DofManager::computePermutation( FieldDescription & field,
+                                     array1d< localIndex > & permutation )
+{
+  localIndex const fieldIndex = getFieldIndex( field.name );
+
+  // step 1: save the number of components, and then set it to 1 temporarily
+  //         do not forget to restore at the end
+  //         we set the number of components to 1 to compute the reordering on a smaller matrix
+
+  localIndex const numComps = field.numComponents;
+  CompMask const globallyCoupledComps = field.globallyCoupledComponents;
+  field.numComponents = 1;
+  field.globallyCoupledComponents = CompMask( 1, true );
+
+  // step 2: compute field dimensions (local dofs, global dofs, etc)
+  //         this is needed to make sure that the sparsity pattern function work properly
+  //         in particular, this function defines the rankOffset (computed with number of components = 1)
+
+  computeFieldDimensions( fieldIndex );
+
+  // the number of local dofs is available at this point, we allocate space for the permutation
+  permutation.resize( numLocalDofs( field.name ) );
+  forAll< parallelHostPolicy >( permutation.size(), [&]( localIndex const i )
+  {
+    permutation[i] = i;
+  } );
+
+  // if no reordering is requesting, we just return the identity permutation
+  if( field.reorderingType == LocalReorderingType::None )
+  {
+    // restore the number of components
+    field.numComponents = numComps;
+    field.globallyCoupledComponents = globallyCoupledComps;
+    // reset the offsets, since they will be recomputed with the proper number of components
+    // this is important to get the right reordering for multiphysics problems
+    field.numLocalDof  = 0;
+    field.rankOffset   = 0;
+    field.numGlobalDof = 0;
+    field.globalOffset = 0;
+    if( fieldIndex > 0 )
+    {
+      field.blockOffset = 0;
+    }
+    return;
+  }
```

Comment:
I like early exit in functions (as opposed to nesting everything else in the `else` branch), but it'd also be nice to not repeat the whole "reset field offsets" block twice. Maybe this function could be refactored a bit by:
1. Isolating `forAll` loop for filling a "trivial" permutation into its own (free or member) function, and RCM-specific code into its own (member) function.
2. `computePermutation()` keeps all common "setup" code (up to and including `computeFieldDimensions()` + allocate `permutation`), followed by a `switch` on `field.reorderingType`, followed by common "teardown" code (reset `field` and return `permutation`).

Reply:
Done, I split the function as you suggested